memory-management/src/dynamic_layout.rs

```diff
@@ -0,0 +1,195 @@
+    }
+
+    pub fn end_offset(&self) -> usize {
+        self.offset_at_index(self.element_count as usize)
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.element_count == 0
+    }
+
+    pub fn len(&self) -> usize {
+        self.element_count as usize
+    }
+
+    pub fn get(&self, element_index: usize) -> Option<&T> {
+        if element_index >= self.element_count as usize {
+            return None;
+        }
+        Some(unsafe {
+            &*((self as *const _ as *const u8).add(self.offset_at_index(element_index)) as *const T)
+        })
+    }
+
+    pub fn get_mut(&mut self, element_index: usize) -> Option<&mut T> {
+        if element_index >= self.element_count as usize {
+            return None;
+        }
+        Some(unsafe {
+            &mut *((self as *mut _ as *mut u8).add(self.offset_at_index(element_index)) as *mut T)
+        })
+    }
+
+    pub fn as_slice(&self) -> &[T] {
+        assert_eq!(self.values_stride as usize, std::mem::size_of::<T>());
+        unsafe {
+            std::slice::from_raw_parts(
+                if self.element_count == 0 {
+                    std::ptr::null()
+                } else {
+                    self.get(0).unwrap() as *const T
+                },
+                self.element_count as usize,
+            )
+        }
+    }
+
+    pub fn as_mut_slice(&mut self) -> &mut [T] {
+        assert_eq!(self.values_stride as usize, std::mem::size_of::<T>());
+        unsafe {
+            std::slice::from_raw_parts_mut(
+                if self.element_count == 0 {
+                    std::ptr::null_mut()
+                } else {
+                    self.get_mut(0).unwrap() as *mut T
+                },
+                self.element_count as usize,
+            )
+        }
+    }
+
+    pub fn iter<'b>(&'b self) -> DynamicLayoutArrayIterator<'a, 'b, T> {
+        DynamicLayoutArrayIterator {
+            _element_type: std::marker::PhantomData::default(),
+            array: self,
+            element_index: 0,
+        }
+    }
+
+    pub fn iter_mut<'b>(&'b mut self) -> DynamicLayoutArrayIteratorMut<'a, 'b, T> {
+        DynamicLayoutArrayIteratorMut {
+            _element_type: std::marker::PhantomData::default(),
+            array: self,
+            element_index: 0,
+        }
+    }
+}
+
+impl<'a, T: std::fmt::Debug> std::fmt::Debug for DynamicLayoutArray<'a, T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        f.debug_list().entries(self.iter()).finish()
+    }
+}
+
+#[derive(Clone)]
+pub struct DynamicLayoutArrayIterator<'a, 'b, T> {
+    _element_type: std::marker::PhantomData<T>,
+    array: &'b DynamicLayoutArray<'a, T>,
+    element_index: usize,
+}
+
+impl<'a, 'b, T> Iterator for DynamicLayoutArrayIterator<'a, 'b, T> {
+    type Item = &'b T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let result = self.array.get(self.element_index);
+        self.element_index = self.element_index.saturating_add(1);
```

Comment:
I think you need to use checked_add here, or for unlikely but technically possible large arrays, the iterator will never finish

Reply:
Not worth it, I just restricted the element count to be one less than the max. Otherwise, this edge case would take even more special handling.